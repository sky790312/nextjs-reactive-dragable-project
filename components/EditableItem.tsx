import React, { useState, useEffect, useRef } from 'react'
import { useDoubleClick } from '@/hooks/useDoubleClick'
import styled from 'styled-components'

interface Props {
  text: string
  shouldHighlight?: boolean
  handleClick: () => void
  handleDoubleClick?: () => void
  handleEditConfirm: (text: string) => void
}

const EditableItem: React.FC<Props> = React.memo(
  ({
    text,
    shouldHighlight = false,
    handleClick,
    handleDoubleClick,
    handleEditConfirm,
  }) => {
    const inputRef = useRef(null)
    const [isEditing, setIsEditing] = useState(false)

    const onClick = () => {
      if (isEditing) {
        return
      }

      handleClick()
    }

    const onDoubleClick = () => {
      setIsEditing(true)
      handleDoubleClick && handleDoubleClick()
    }

    const onItemClick = useDoubleClick(
      () => onClick(),
      () => onDoubleClick(),
    )

    const onEditConfirm = () => {
      setIsEditing(false)
      handleEditConfirm(inputRef.current.value)
    }

    useEffect(() => {
      if (isEditing) {
        inputRef.current.focus()
      }
    }, [isEditing])

    return (
      <div onClick={onItemClick}>
        {isEditing ? (
          <>
            <input type='text' defaultValue={text} ref={inputRef} />
            <button onClick={onEditConfirm}>ok</button>
          </>
        ) : (
          <Text shouldHighlight={shouldHighlight}>{text}</Text>
        )}
      </div>
    )
  },
)

export default EditableItem

const Text = styled.div<{ shouldHighlight: boolean }>`
  ${props => {
    if (props.shouldHighlight) {
      return `
        cursor: default;
        font-weight: bold;
      `
    } else {
      return `
        cursor: pointer;
      `
    }
  }}
`
