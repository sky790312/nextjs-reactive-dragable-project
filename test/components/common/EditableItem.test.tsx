import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import EditableItem from '@/components/EditableItem'

describe('EditableItem component', () => {
  it('matches snapshot', () => {
    const props = {
      text: '',
      handleClick: jest.fn(),
      handleEditConfirm: jest.fn(),
    }
    const { asFragment } = render(
      withProviders(<EditableItem {...props} />),
      {},
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
