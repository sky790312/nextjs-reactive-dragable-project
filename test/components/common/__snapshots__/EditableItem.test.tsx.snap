// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableItem component matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="EditableItem__Text-apsj0b-0 jUHTNR"
    />
  </div>
</DocumentFragment>
`;
